libraries/app/api.cpp

```diff
@@ -134,23 +134,65 @@ namespace graphene { namespace app {
 
     void network_broadcast_api::on_applied_block( const signed_block& b )
     {
-       if( _callbacks.size() )
-       {
-          /// we need to ensure the database_api is not deleted for the life of the async operation
-          auto capture_this = shared_from_this();
-          for( uint32_t trx_num = 0; trx_num < b.transactions.size(); ++trx_num )
+       /// we need to ensure the database_api is not deleted for the life of the async operation
+       auto capture_this = shared_from_this();
+
+       fc::async( [this,capture_this,b]() {
+          int32_t block_num = int32_t(b.block_num());
+          if( _callbacks.size() )
           {
-             const auto& trx = b.transactions[trx_num];
-             auto id = trx.id();
-             auto itr = _callbacks.find(id);
-             if( itr != _callbacks.end() )
+             for( int32_t trx_num = 0; trx_num < b.transactions.size(); ++trx_num )
              {
-                auto block_num = b.block_num();
-                auto& callback = _callbacks.find(id)->second;
-                fc::async( [capture_this,this,id,block_num,trx_num,trx,callback](){ callback( fc::variant(transaction_confirmation{ id, block_num, trx_num, trx}) ); } );
+                const auto& trx = b.transactions[trx_num];
+                auto id = trx.id();
+                auto itr = _callbacks.find(id);
+                if( itr == _callbacks.end() ) continue;
+                confirmation_callback callback = itr->second;
+                itr->second = [](variant){};
+                callback( fc::variant(transaction_confirmation{ id, block_num, trx_num, false }) );
+                _callbacks.erase( itr );
```

Comment:
Could the transaction id be erased from the vector in _callbacks_expirations here? If so, we should probably also remove the time_point_sec entry if the vector is empty.